TypeScript support: don't try to compile templates

benmccann opened the pull request:
Enables TypeScript support https://github.com/sveltejs/sapper/issues/760

TypeScript won't compile with Svelte compiler and first needs to be preprocessed. However, that happens after this check because `create_app` outputs JS which imports these `preload` exports before we try to bundle the whole app

This also has to be much faster to build the mainfest!

thgh commented:
Maybe add an extra check for the file extension and only disable for `.ts`?

There is a usecase where you may want to compose the preload function:
`export const preload = verifyUser(catchErrors(function () { ... }))`

Also reusable preload functions:
`export { preload } from './userData'`

benmccann commented:
Thanks @thgh !  I've updated the regex to cover both the cases you mentioned. Take another look and let me know what you think

> Maybe add an extra check for the file extension and only disable for .ts?

The code is typically in a `.svelte` file, so we couldn't do that unfortunately. But I think this will be a really nice speed up for building apps in JS anyway.

benmccann commented:
I've been using this in my fork for the past few weeks and the only issue I ran into was if `preload` was commented out. I've updated this PR to address that case and added several new unit tests to cover all the cases that have been suggested so far.

Conduitry commented:
Another possibility might be https://github.com/Rich-Harris/tippex

benmccann commented:
Cool. I'm happy to switch it out if you prefer. `strip-comments` has been working well though and is a heavily used library

Conduitry commented:
Well tippex seems to match the case we have here better because it also blanks out strings from its input, which strip-comments does not but which is needed to help avoid false positives.

benmccann commented:
Ah, yeah that make sense. I tried to use Tippex, but unfortunately it seems to have a bug which prevents it from being used here https://github.com/Rich-Harris/tippex/issues/23. So it's probably best to stick with `strip-comments` for now and if the bug is fixed we can switch out `strip-comments` for `tippex` then

benmccann commented:
Using Tippex is substantially more complicated. I can run `strip-comments` on the whole file. With Tippex, I need to first identify the location of each `script` block in the file and then extract its contents. Is there some existing piece of code in the Svelte compiler or preprocessor that can be reused for that? Or are we worried enough about encountering something like `let func: string = 'export async function preload()';` that it's worth it?

p.s. the test failures are unrelated to this PR and are coming from https://github.com/sveltejs/sapper/pull/743 having been merged

PatrickG commented:
In the meantime, maybe we can only compile the module script block, until this is resolved?

Like this:
```js
let matches;
if (/preload/.test(source) && (matches = source.match(/<script context="module">[^]*?<\/script>/))) {
  try {
    const { vars } = svelte.compile(matches[0], { generate: false });
    return vars.some((variable: any) => variable.module && variable.export_name === 'preload');
  } catch (err) {}
}
```

The benefit would be, to use typescript in the other script block.

benmccann commented:
I've gone ahead and switched this to use https://github.com/Rich-Harris/tippex

benmccann commented:
@adamb pointed out to me over Discord that there were some problems using `tippex`, so I've switched this back to using `strip-comments`. The issue with `tippex` is that it requires just the `script` contents to be passed in. However, it's basically impossible to get the `script` contents in a robust manner without writing a full-blown parser. E.g. consider the following:

```
<script context="module">
const x = `<script>...</script>`
export function preload() { }
</script>
```

We have a few options here:
- Sometimes hit a false positive on detecting `preload` with a simple regex. That's this PR
  - This seems acceptable to me because I think it's rare enough. The only case where I can realistically see it happening is if someone is writing about Sapper in `.svelte` files in Sapper. I checked the Sapper documentation and it would not be affected because it's primarily written in `.md`. It also wouldn't affect any users who are loading contents from any dynamic source like a database
- Sometimes hit a false negative on detecting `preload` by trying to extract the `script` contents and then use Tippex like in the above example
  - I think we could make the logic for extracting the script contents more robust so that failures were very unlikely, but we'd still have to accept that it would not always be perfect. I think that we could make those cases rare enough that it'd be a fine way forward though. E.g. we'd probably have to accept not handling this case:
```
<script context="module">
let x = y ? `<script context="module">var foobar="y";` : `<script context="module">var foobar="x";`;
x += '</script>';
export function preload() { }
</script>
```
- Try to first detect if the project is using TypeScript. Use the Svelte compiler to check if it's valid JavaScript just as we were doing before if the project is not a TypeScript project. Use the above logic only for TypeScript projects
  - Probably the best way to do this is just have the user specify if they are using TypeScript in the Sapper config file. The other ways I can think of also seem error-prone. I'd really like not to block on the config RFC though, so hope that we could either use one of the unperfect ways of detecting TypeScript in the meantime or leave this as an enhancement to implement later
- Don't support TypeScript
  - My least favorite option. TypeScript is hugely popular
- Massively refactor the entire Sapper project to be a series of Rollup plugins so that the preprocessor can executed before `create_manifest_data` and `create_app`
  - This would be so challenging that it's basically equivalent to the previous bullet. I'm not entirely sure it's even feasible, but if it were it would be a huge amount of work
  - I think it's somewhat unnecessary to be this robust because I can also come up with cases where the code currently checked into `master` does not work and it hasn't been an issue. I think it's acceptable to accept some edge case if it's rare enough. E.g. the following would break today:
```
<script context="module">
const x = `<style>`
export function preload() { }
</script>

Hello world
<!--
</style>
-->
```

benmccann commented:
The main thing I was keeping my eyes on with this PR is the fact that it uses a regex to extract the `<script context="module">` tags. It turns out that svelte itself uses a regex to extract `script` tags, so I think it's highly unlikely that this would cause any regression. If Svelte comes up with any other way of doing this in the future then we can always update Sapper to match

https://github.com/sveltejs/svelte/blob/40987b7780ef1865fb94e89b4d2f5a53a71056f9/src/compiler/preprocess/index.ts#L97

Conduitry commented:
This was a good idea, but closing in favor of #1344, which is a better idea.
